doc/port-fec/fec-auto-determination.md

```diff
@@ -0,0 +1,187 @@
+- [API design](#api-design)
+- [Common Rule for FEC Determination](#common-rule-for-fec-determination)
+    - [Table 1: FEC Mapping Based on Optics Type](#table-1-fec-mapping-based-on-optics-type)
+    - [Table 2: FEC Mapping Based on Lane Speed and Number of Lanes](#table-2-fec-mapping-based-on-lane-speed-and-number-of-lanes)
+- [Diagram For Different Use Cases](#diagram-for-different-use-cases)
+- [Dependency](#dependency)
+
+### Revision
+
+ | Rev |     Date    |       Author        | Change Description                         |
+ |:---:|:-----------:|:-------------------:|--------------------------------------------|
+ | 0.1 |             |     Shyam Kumar, Longyin Huang   | Initial version                            |
+
+### Scope
+This document is the design document for FEC auto-determination feature on SONiC.
+
+### Definitions/Abbreviations
+| **Term**       | **Definition**                                   |
+| -------------- | ------------------------------------------------ |
+| FEC            | Forward Error Correction                         |
+| DPB            | Dynamic Port Breakout                            |
+
+### Overview
+
+FEC mode is a critical configuration for a port, which needs to be configured properly for the port to come up.
+
+There are below scenarios that can end up with wrong FEC mode:
+1. In DPB(Dynamic Port Breakout) case, today's DPB CLI doesn't generate FEC config for newly created ports, FEC mode is default to `none` at [SAI/SDK](https://github.com/opencomputeproject/SAI/blob/a94bbbe43242a4d9e1a4d9f70780ea9251127f5d/inc/saiport.h#L1012) layer.
+2. In non-DPB case,
+    - Some platforms have no FEC configured in CONFIG_DB by default. The FEC mode can be either default to `none` at SAI/SDK layer or manually configured by user who might not have enough domain knowledge.
+    - Some platforms have default FEC defined in `port_config.ini`, which however might not be suitable for the specific port/optics on the system.
+
+The feature in this document is to address the issue in both of above scenarios in a common platform-independent way, since the rule to determine FEC for a given port/optics is common for all platforms.
+
+### High-Level Design
+
+Add `determine-fec` module which can determine FEC mode based on common rule for a given port with a given optics. This module provides a `determine_fec` API which can be invoked in below use cases:
+1. DPB use case: Enhance today's DPB CLI to automatically determine and configure FEC in CONFIG_DB for dynamically created ports, based on determine-fec module.
+2. non-DPB use case: Add a user-triggered CLI `fec-auto-correct` to automatically determine and configure FEC in CONFIG_DB for existing ports, based on determine-fec module.
+    - Future plan: determine-fec module can be further enhanced to be integrated with xcvrd, which can be triggered automatically during transceiver insertion, without human intervention. (details TBD)
```

Comment:
Please update the HLD as discussed in the community to capture the flow where FEC setting is generated by xcvrd based on media type and pushed to application DB.